Sources/MongoSwift/BSON/BSONValue.swift

```diff
@@ -732,25 +733,40 @@ public struct ObjectId: BSONValue, Equatable, CustomStringConvertible, Codable {
     public var bsonType: BSONType { return .objectId }
 
     /// This `ObjectId`'s data represented as a `String`.
-    public let oid: String
+    public var hex: String {
+        var str = Data(count: 25)
+        return str.withUnsafeMutableBytes { (rawBuffer: UnsafeMutablePointer<Int8>) in
+            withUnsafePointer(to: self.oid) { oidPtr in
+                bson_oid_to_string(oidPtr, rawBuffer)
+            }
+            return String(cString: rawBuffer)
+        }
+    }
 
     /// The timestamp used to create this `ObjectId`
-    public let timestamp: UInt32
+    public var timestamp: UInt32 {
+        return withUnsafePointer(to: self.oid) { oidPtr in UInt32(bson_oid_get_time_t(oidPtr)) }
+    }
+
+    public var description: String {
+        return self.hex
+    }
+
+    internal let oid: bson_oid_t
 
     /// Initializes a new `ObjectId`.
     public init() {
-        var oid_t = bson_oid_t()
-        bson_oid_init(&oid_t, nil)
-        self.init(fromPointer: &oid_t)
+        var oid = bson_oid_t()
+        bson_oid_init(&oid, nil)
+        self.oid = oid
     }
 
     /// Initializes an `ObjectId` from the provided `String`. Assumes that the given string is a valid ObjectId.
     /// - SeeAlso: https://github.com/mongodb/specifications/blob/master/source/objectid.rst
     public init(fromString oid: String) {
```

Comment:
any opinions on whether we should keep this parameter label? I feel like just `from` is more in line with the rest of the driver, and the `String` isn't needed to disambiguate from the decodable initializer.

Reply:
I agree `fromString` should probably be replaced. In fact, I kind of think the entire initializer should be removed, and `init?(ifValid: String)` should be the only one, with the `ifValid` label being replaced with either `from` or `hex` or `fromHex`. 

The libbson docs do not mention what happens if `bson_oid_init_from_string` is called with an invalid string, so I guess we can say it's undefined behavior. I really think we should minimize the amount of UB easily available through our API if at all possible. Thoughts?

Reply:
I like the idea of having a single failable initializer. that is consistent with how we handle `Decimal128`s.

Reply:
Refactored `init?(ifValid: String)` to just `init?(_:  String)`. Let me know if you think it'd be better with a label. Also, removed the un-failable init.

Reply:
I could go either way.... it might be useful as an indicator about the type of string, but maybe that's not that necessary.

Reply:
@mbroadst any strong opinions on this one? I feel like I could go either way too. The one benefit of doing it without a label is that it looks just like it would in the shell, though I doubt we generally make many driver design choices based on how the shell behaves.